s2fft/precompute_transforms/custom_ops.py

```diff
@@ -0,0 +1,410 @@
+
+    """
+    if sampling.lower() not in ["mw", "mwss"]:
+        raise ValueError(
+            f"Fourier-Wigner algorithm does not support {sampling} sampling."
+        )
+
+    # EXTRACT VARIOUS PRECOMPUTES
+    Delta, _ = DW
+
+    # INDEX VALUES
+    n_dim = len(spins)
+    m_offset = 1 if sampling.lower() == "mwss" else 0
+    ntheta = L + 1 if sampling.lower() == "mwss" else L
+    theta0 = 0 if sampling.lower() == "mwss" else jnp.pi / (2 * L - 1)
+    xnlm_size = 2 * L if sampling.lower() == "mwss" else 2 * L - 1
+
+    # REUSED ARRAYS
+    m = jnp.arange(-L + 1 - m_offset, L)
+    n = -spins
+
+    # Calculate fmna = i^(n-m)\sum_L Delta^l_am Delta^l_an f^l_mn(2l+1)/(8pi^2)
+    x = jnp.zeros(
+        (flmn.shape[0], n_dim, xnlm_size, xnlm_size, flmn.shape[-1]), dtype=flmn.dtype
+    )
+    x = x.at[:, :, m_offset:, m_offset:, :].set(
+        jnp.einsum(
+            "bnlmc,lam,lan,l->bnamc",
+            flmn,
+            Delta,
+            Delta[:, :, L - 1 + n],
+            (2 * jnp.arange(L) + 1) / (8 * jnp.pi**2),
+        )
+    )
+
+    # APPLY SIGN FUNCTION AND PHASE SHIFT
+    x = jnp.einsum(
+        "bnamc,m,n,a->bnamc", x, 1j ** (-m), 1j ** (n), jnp.exp(1j * m * theta0)
+    )
+
+    # IFFT OVER THETA AND PHI
+    x = jnp.fft.ifftshift(x, axes=(-3, -2))
+    x = jnp.fft.ifft(x, axis=-3, norm="forward")[:, :, :ntheta, :, :]
+    return jnp.fft.ifft(x, axis=-2, norm="forward")
+
+
+def so3_to_wigner_subset(
+    f: np.ndarray,
+    spins: np.ndarray,
+    DW: Tuple[np.ndarray, np.ndarray],
+    L: int,
+    N: int,
+    sampling: str = "mw",
+) -> np.ndarray:
+    r"""
+    Transforms a signal on the rotation group to an arbitrary subset of its Wigner
+    coefficients.
+
+        This function takes a signal on the rotation group SO(3) and computes a subset of
+        spin spherical harmonic coefficients corresponding to slices across the requested
+        spin numbers. These spin numbers can be arbitrarily chosen such that their absolute
+        value is less than or equal to the azimuthal band-limit :math:`N\leq L`.
+
+    Args:
+        f (np.ndarray): Signal on the rotation group with shape :math:`[batch, n_\gamma, n_\theta,n_\phi, channels]`.
+        spins (np.ndarray): Spins of each field in fs with shape :math:`[n_s]`.
+        DW (Tuple[np.ndarray, np.ndarray]): Fourier coefficients of the reduced
+            Wigner d-functions and the corresponding upsampled quadrature weights.
+        L (int): Harmonic band-limit.
+        N (int): Azimuthal band-limit.
+        sampling (str, optional): Sampling scheme.  Supported sampling schemes include
+            {"mw", "mwss"}. Defaults to "mw".
+
+    Returns:
+        np.ndarray: Collection of spin spherical harmonic coefficients with shape :math:`[batch, n_s, L, 2L-1, channels]`.
+
+    """
+    # COMPUTE FFT OVER GAMMA
+    x = np.fft.fft(f, axis=-4, norm="forward")
```

Comment:
Shouldn't we do a DFT here rather than FFT so we only compute the spins of interest?  Rather than computing all and selecting a subset.

Reply:
Similiarly for the jax implementation below?

Reply:
Yep, good catch. I've switched this out to an explicitly DFT.